### PR TITLE
feat(http): Force default rate limits for some known hosts

### LINKED ISSUE #27802: Apply default rate limits for hosts/registries with known rate limit needs
### Describe the proposed change(s).

For example, the npmjs registry or GOPROXY does not need rate limiting applied, while the following are known to need:
- Repology
- crates.io
- Rubygems API

Let's identify all such hosts in a single file/location, in a file like `lib/data/host-limits.ts`. I think we can reuse `matchHost` syntax, it can maybe even be `hostRules` format but is not necessary (the `matchHost` value can be a key for the Record).

Importantly, if a host has a default limit in this file then we should make sure that users cannot _increase_ the rate limit using `hostRules`. Ideally we can catch and warn if such an attempt is made. This protects fragile registries from falling victim to flooding from Renovate users unless they really go out of their way to overcome this.

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Default limits can be overriden only towards more restriction
- 8 requests per second for `rubygems.org` 
- 1 request per second for `https://crates.io/api`
- 16 concurrent requests for each host by default

## Context

- Ref: #27369
- Closes: #27802

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by HonkingGoose — COMMENTED** [review]
Do you have a source for the new rate limit numbers?

Maybe the project has documentation that says: "please limit to `x` requests per `y` time"?

File: lib/util/http/queue.ts
Comment (viceice) [inline lib/util/http/queue.ts:23]:
```suggestion
      ...[hostConcurrency, defaultConcurrency].filter(is.notNull),
```

File: lib/util/http/throttle.ts
Comment (viceice) [inline lib/util/http/throttle.ts:40]:
```suggestion
      ...[hostThrottleMs, defaultThrottleMs].filter(is.notNull),
```

**Review by viceice — COMMENTED** [review]
otherwise LGTM

File: lib/util/http/rate-limit.spec.ts
Comment (HonkingGoose) [inline lib/util/http/rate-limit.spec.ts:31]:
```suggestion
    it('selects default value if host rule is greater', () => {
```

File: lib/util/http/rate-limit.spec.ts
Comment (HonkingGoose) [inline lib/util/http/rate-limit.spec.ts:61]:
```suggestion
    it('gets the limit from the host rules', () => {
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by HonkingGoose — APPROVED** [review]
I'm happy with the test descriptions now. I'll let the maintainers review the rest.

File: lib/util/http/rate-limits.ts
Comment (viceice) [inline lib/util/http/rate-limits.ts:19]: @rarkins do we really want this? 🤔

File: lib/util/http/rate-limits.ts
Comment (viceice) [inline lib/util/http/rate-limits.ts:27]: this is not fail save against runtime null passing

Comment (zharinov) [reply lib/util/http/rate-limits.ts:27]: In the runtime, it's meant to be called with no arguments, hence set `defaults`.
For tests, it remains empty everywhere except `rate-limit.spec.ts`.

What would you like to change?

Comment (viceice) [reply lib/util/http/rate-limits.ts:27]:
```suggestion
export function setHttpRateLimits(rules?: RateLimitRule[]): void {
  limits = rules ?? defaults;
```
so it's null and undefined safe for any invalid code ussge

File: lib/util/http/rate-limits.ts
Comment (rarkins) [reply lib/util/http/rate-limits.ts:19]: Probably a good idea. For example npm v9 had a default of 15 concurrent requests, which was lost in v10, and apparently it caused some hosts load problems.

**Comment by zharinov** [comment]
This PR seems to be the only blocker for #27369

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**
